userspace/libsinsp/sinsp_cycledumper.cpp

```diff
@@ -0,0 +1,251 @@
+// SPDX-License-Identifier: Apache-2.0
+/*
+Copyright (C) 2023 The Falco Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+*/
+
+#include "sinsp_cycledumper.h"
+#include <iostream>
+
+
+sinsp_cycledumper::sinsp_cycledumper(sinsp* inspector, const std::string& base_filename,
+                    const int& rollover_mb, const int& duration_seconds,
+                    const int& file_limit, const unsigned long& event_limit,
+                    const bool& compress):
+	m_last_time(0),
+	m_file_count_total(0),
+	m_file_index(0),
+	m_has_started(false),
+	m_event_count(0L),
+	m_past_names(NULL),
+	m_limit_format("")
+{
+	m_base_filename = base_filename;
+	m_rollover_mb = rollover_mb * 1000000L;
+	m_duration_seconds = duration_seconds;
+	m_file_limit = file_limit;
+	m_event_limit = event_limit;
+    m_inspector = inspector;
+    m_compress  = compress;
+
+	if(duration_seconds > 0 && file_limit > 0)
+	{
+		m_past_names = new std::string[file_limit];
+
+		for(int32_t j = 0; j < file_limit; j++)
+		{
+			m_past_names[j] = "";
+		}
+	}
+}
+
+sinsp_cycledumper::~sinsp_cycledumper()
+{
+	if(m_dumper != nullptr)
+	{
+		m_dumper->close();
+		m_dumper.reset(nullptr);
+	}
+}
+
+void sinsp_cycledumper::dump(sinsp_evt* evt)
+{
+    if(is_new_file_needed(evt))
+	{
+		autodump_next_file();
+    }
+    m_dumper->dump(evt);
+}
+
+void sinsp_cycledumper::close()
+{
+	autodump_stop();
+}
+
+void sinsp_cycledumper::autodump_next_file()
+{
+	autodump_stop();
+	autodump_start(m_current_filename);
+}
+
+void sinsp_cycledumper::autodump_stop()
+{
+	if(!m_inspector)
+	{
+		throw sinsp_exception("inspector not opened yet");
+	}
+
+	if(m_dumper)
+	{
+		m_dumper->close();
+		m_dumper = NULL;
+	}
+
+	m_inspector->m_is_dumping = false;
+}
+
+void sinsp_cycledumper::autodump_start(const std::string& dump_filename)
+{
+	if(!m_inspector)
+	{
+		throw sinsp_exception("inspector not opened yet");
+	}
+
+	m_dumper.reset(new sinsp_dumper);
```

Comment:
Can we consistently use smart pointers instead?

Reply:
Yeap! Absolutely! :)